### PR TITLE
create-user-collection

### PR DESCRIPTION
**Pull request by asenhodzhev**
Title: Create user collection.

Description: User collection is created with all crud operations.

Ticket link: [create-user-collection](https://github.com/Sf-62-NodeJS/clothes-market-server/issues/7)
[extend-collection-user-functionality](https://github.com/Sf-62-NodeJS/clothes-market-server/issues/40)

### CODE REVIEW

File: src/controllers/users.js [hunk 1]
```diff
@@ -12,6 +15,70 @@ class UsersController {
 
     return res.json(newUser);
   }
+
+  async deleteUser (req, res) {
+    try {
+      await User.deleteOne({ _id: req.params.id }).exec();
+
+      return res.status(200).json({ message: 'User is deleted.' });
+    } catch (error) {
```
Comment (VasylLymych) [inline src/controllers/users.js]: Return always status 200 (in case if there is no error during connection to db) and true/false in case user has been deleted/not deleted

File: src/controllers/users.js [hunk 2]
```diff
@@ -12,6 +15,70 @@ class UsersController {
...
+      res.status(400).json({ err: error });
```
Comment (VasylLymych) [inline src/controllers/users.js]: Use res.boom instead status

File: src/controllers/users.js [hunk 3]
```diff
@@ -12,6 +15,70 @@ class UsersController {
...
+    }
+  }
+
+  async getUsers (req, res) {
+    const { _id, name, email, status, role, page, docs } = req.body;
```
Comment (VasylLymych) [inline src/controllers/users.js]: All these params will be in query params (req.query). Name them skip and take instead of page and docs

File: src/controllers/users.js [hunk 4]
```diff
@@ -12,6 +15,70 @@ class UsersController {
...
+    if (name != null) query.name = name;
+    if (email != null) query.email = email;
+    if (status != null) query.status = status;
+    if (role != null) query.role = role;
+    const pageNumber = page || PAGE_NUMBER;
+    const docsNumber = docs || DOCS_NUMBER;
+
+    try {
+      const users = await User.find(query)
+        .skip((pageNumber - 1) * docsNumber)
```
Comment (VasylLymych) [inline src/controllers/users.js]: All these counts must be performed on client side. In our case server must accept skip/take

File: src/controllers/users.js [hunk 5]
```diff
@@ -12,6 +15,70 @@ class UsersController {
...
+        .limit(docsNumber)
+        .exec();
+
+      return users ? res.json(users) : res.boom.notFound();
```
Comment (VasylLymych) [inline src/controllers/users.js]: Return an empty array if there are no users instead of error

File: src/controllers/users.js [hunk 6]
```diff
@@ -12,6 +15,70 @@ class UsersController {
...
+
+  async getUsers (req, res) {
+    const { _id, name, email, status, role, page, docs } = req.body;
+    const query = {};
+    if (_id != null) query._id = _id;
+    if (name != null) query.name = name;
+    if (email != null) query.email = email;
+    if (status != null) query.status = status;
+    if (role != null) query.role = role;
+    const pageNumber = page || PAGE_NUMBER;
```
Comment (VasylLymych) [inline src/controllers/users.js]: No necessity to create separate variables PAGE_NUMBER... Just write inline +skip || 0, +take || 50

File: src/controllers/users.js [hunk 7]
```diff
@@ -12,6 +15,70 @@ class UsersController {
...
+
+    try {
+      const users = await User.find(query)
+        .skip((pageNumber - 1) * docsNumber)
+        .limit(docsNumber)
+        .exec();
+
+      return users ? res.json(users) : res.boom.notFound();
+    } catch (error) {
+      res.status(400).json({ err: error });
```
Comment (VasylLymych) [inline src/controllers/users.js]: res.boom

File: src/controllers/users.js [hunk 8]
```diff
@@ -12,6 +15,70 @@ class UsersController {
...
+    }
+  }
+
+  async blockUser (req, res) {
```
Comment (VasylLymych) [inline src/controllers/users.js]: Actually we don't need here method block user. We have method update user, so we must be able to block user there

File: src/controllers/users.js [hunk 9]
```diff
@@ -12,6 +15,70 @@ class UsersController {
...
+      return res.status(200).json({ message: 'User is blocked.' });
+    } catch (error) {
+      res.status(400).json({ err: error });
+    }
+  }
+
+  async updateUser (req, res) {
+    const { name, surname, middleName, email, phoneNumber } = req.body;
+    try {
+      const updatedUser = await User.findByIdAndUpdate(
```
Comment (VasylLymych) [inline src/controllers/users.js]: Check if user exists...

File: src/controllers/users.js [hunk 10]
```diff
@@ -12,6 +15,70 @@ class UsersController {
...
+          surname,
+          middleName,
+          email,
+          phoneNumber
+        },
+        { new: true }
+      ).exec();
+      res.status(200).json(updatedUser);
+    } catch (error) {
+      res.status(400).json({ err: error });
```
Comment (VasylLymych) [inline src/controllers/users.js]: res.boom

File: src/index.js [hunk 11]
```diff
@@ -33,25 +38,34 @@ app.use(boom());
...
+  swaggerUi.setup(swagger, { explorer: true })
+);
 expressWinston.requestWhitelist.push('body');
 app.use(expressWinston.logger(loggerOptions));
 app.use(expressWinston.errorLogger(loggerOptions));
 
 // DB connect
-mongoose.connect(process.env.MONGO_URL);
+// mongoose.connect(process.env.MONGO_URL);
+mongoose.connect('mongodb://127.0.0.1:27017/test', { useNewUrlParser: true });
```
Comment (VasylLymych) [inline src/index.js]: Why did you add this?

File: src/index.js [hunk 12]
```diff
@@ -33,25 +38,34 @@ app.use(boom());
...
 
 // Routes
 app.use('/users', usersRouter);
 app.disable('etag');
 
-app.listen(process.env.PORT, (err) => {
-  if (err) {
-    console.log(err.message);
-  }
+if (process.env.NODE_ENV !== 'test') {
```
Comment (VasylLymych) [inline src/index.js]: Our app must not care about tests and test environment. Why have you added this check?

File: src/middlewares/validators/createUserPayload.js [hunk 13]
```diff
@@ -3,9 +3,40 @@ const Joi = require('joi');
...
+
+    middleName: Joi.string()
+      .alphanum()
+      .min(2)
+      .max(30)
+      .trim(true)
+      .pattern(/^[a-zA-Z0-9]{3,30}$/)
+      .required(),
+
+    hashedPassword: Joi.string().required(),
```
Comment (VasylLymych) [inline src/middlewares/validators/createUserPayload.js]: User pass as payload simple password, not hashed

File: src/middlewares/validators/createUserPayload.js [hunk 14]
```diff
@@ -3,9 +3,40 @@ const Joi = require('joi');
...
+
+    phoneNumber: Joi.string()
+      .trim(true)
+      .length(10)
+      .pattern(/^[0-9]+$/)
+      .required(),
+
+    address: Joi.string().trim(true).lowercase().required(),
```
Comment (VasylLymych) [inline src/middlewares/validators/createUserPayload.js]: lowercase excessive here

File: src/middlewares/validators/createUserPayload.js [hunk 15]
```diff
@@ -3,9 +3,40 @@ const Joi = require('joi');
...
+
+    email: Joi.string().trim(true).lowercase().email().required()
```
Comment (VasylLymych) [inline src/middlewares/validators/createUserPayload.js]: lowercase excessive here

File: src/middlewares/validators/createUserPayload.js [hunk 16]
```diff
@@ -3,9 +3,40 @@ const Joi = require('joi');
...
       .max(30)
-      .required()
+      .trim(true)
+      .pattern(/^[a-zA-Z0-9]{3,30}$/)
+      .required(),
+
+    surname: Joi.string()
+      .alphanum()
+      .min(2)
+      .max(30)
```
Comment (VasylLymych) [inline src/middlewares/validators/createUserPayload.js]: Surname can be longer. Same as name and middlename

File: src/controllers/users.js [hunk 17]
```diff
@@ -1,5 +1,8 @@
 const { User } = require('../models');
 
+const PAGE_NUMBER = 1;
+const DOCS_NUMBER = 50;
+
 class UsersController {
   async getUser (req, res) {
```
Comment (VasylLymych) [inline src/controllers/users.js]: Remove this method. We have method getUsers which contain id filter

File: src/routers/users.js [hunk 18]
```diff
@@ -4,6 +4,10 @@ const { createUserPayloadValidator } = require('../middlewares/validators');
 
 const usersController = new UsersController();
 usersRouter.post('/', createUserPayloadValidator, usersController.createUser);
+usersRouter.post('/all_users', usersController.getUsers);
```
Comment (VasylLymych) [inline src/routers/users.js]: /users, not /all_users

File: src/index.js (hunk 12 above)
Comment (asenhodzhev) [reply src/index.js]: Because when I execute the tests, I am getting this error for createUser():
listen EADDRINUSE: address already in use :::5000

File: src/controllers/tests/users.test.js [hunk 19]
```diff
@@ -27,14 +27,14 @@ describe('usersController tests', () => {
     boom: {
       badRequest: jest.fn()
     },
-    json: payload => payload
+    json: (payload) => payload
   };
 
-  it('should return user', async () => {
+  /* it('should return user', async () => {
```
Comment (VasylLymych) [inline src/controllers/tests/users.test.js]: Remove this test instead comment. Also add tests for all requests: createUser, updateUser, deleteUser, getUsers

File: src/index.js [hunk 20]
```diff
@@ -49,6 +49,7 @@ app.use(expressWinston.logger(loggerOptions));
 app.use(expressWinston.errorLogger(loggerOptions));
 
 // DB connect
+
```
Comment (VasylLymych) [inline src/index.js]: Remove this empty line

File: src/middlewares/validators/createUserPayload.js [hunk 21]
```diff
@@ -2,10 +2,28 @@ const Joi = require('joi');
...
+      .alphanum()
       .min(2)
-      .max(30)
-      .required()
+      .max(45)
+      .trim(true)
+      .required(),
+
+    password: Joi.string().min(8).max(100).required(),
+
```
Comment (VasylLymych) [inline src/middlewares/validators/createUserPayload.js]: Remove all these empty lines, they are excessive here

File: src/routers/users.js [hunk 22]
```diff
@@ -4,6 +4,8 @@ const { createUserPayloadValidator } = require('../middlewares/validators');
 
 const usersController = new UsersController();
 usersRouter.post('/', createUserPayloadValidator, usersController.createUser);
-usersRouter.get('/:id', usersController.getUser);
+usersRouter.put('/:id', usersController.updateUser);
```
Comment (VasylLymych) [inline src/routers/users.js]: We also need middleware "updateUserPayload", because user cannot update all fields (email for example)

**Comment by VasylLymych** [comment]
Describe your endpoints in swagger. Fix also sonar check code smell

**Comment by asenhodzhev** [comment]
post
get
put
delete

File: src/swagger.json [hunk 23]
```diff
@@ -95,35 +219,105 @@
...
               }
             }
           },
           "500": {
             "description": "Internal Server Error"
           }
         }
-      }
+      },
+
```
Comment (VasylLymych) [inline src/swagger.json]: Remove all these empty lines

File: src/services/users.js [hunk 24]
```diff
@@ -0,0 +1,44 @@
+const { User } = require('../models');
+
+class UsersService {
+  async createUser (req, res) {
+    const newUser = await new User(req.body).save();
```
Comment (VasylLymych) [inline src/services/users.js]: Every user must have role "User" by default. Also user must have status "Active" by default

File: src/middlewares/validators/updateUserPayload.js [hunk 25]
```diff
@@ -0,0 +1,21 @@
+const Joi = require('joi');
+
+const updateUserPayloadValidator = (req, res, next) => {
+  const schema = Joi.object({
+    name: Joi.string().alphanum().min(2).max(45).trim(true),
+    surname: Joi.string().alphanum().min(2).max(45).trim(true),
+    middleName: Joi.string().alphanum().min(2).max(45).trim(true),
+    password: Joi.string().min(8).max(100),
```
Comment (VasylLymych) [inline src/middlewares/validators/updateUserPayload.js]: Create please separate endpoint to update password, which accepts oldPassword and newPassword, compares old passwords and then update it. Create please also seperate ticket to add 3 new endpoints: createAdmin (it has the same structure as createUser, but role will be "Admin"), blockUser (accepts user id in req.params and sets userStatus as "Blocked"), deleteUser (accepts user id in req.params and sets userStatus as "Deleted")

**Comment by VasylLymych** [comment]
> post get put delete

Remove password from all of these responses. Return true or false on createUser request. Also password must be hashed. Why is it "password7"?

File: src/services/users.js [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+const { UserRoles } = require('../models');
+const { UserStatuses } = require('../models');
+const bcrypt = require('bcryptjs');
+
+class UsersService {
+  async updateUserPassword (req, res) {
+    const { oldPassword, newPassword } = req.body;
+    const user = await User.findOne({ _id: req.params.id }).exec();
+
+    bcrypt.compare(oldPassword, user.password, async function (err, isMatch) {
```
Comment (vhazna) [inline src/services/users.js:11]: Try to add 'return' before 'bcrypt.compare()' function.

File: src/services/users.js [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+          const salt = await bcrypt.genSalt(10);
+          user.password = await bcrypt.hash(user.password, salt);
+          await user.save();
+          return user ? res.json(true) : res.boom.notFound();
+        } else {
+          res.boom.badRequest();
+        }
+      });
+    } catch (err) {
+      res.boom.badRequest();
```
Comment (VasylLymych) [inline src/services/users.js]: Add some messages to an error. There must an errors like: "new password cannot be the same as old", "wrong old password"

File: src/services/users.js [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+  }
+
+  async getRoleAdmin (req, res) {
+    const roleAdmin = await UserRoles.findOne({ name: 'Admin' });
+    return roleAdmin;
+  }
+
+  async createAdmin (req, res) {
+    try {
+      const roleAdmin = await this.getRoleAdmin();
```
Comment (VasylLymych) [inline src/services/users.js]: Code duplication. Create some mutual method for create User/Admin

File: src/services/users.js [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+
+  async getStatusBlocked (req, res) {
+    const statusBlocked = await UserStatuses.findOne({ name: 'Blocked' });
+    return statusBlocked;
+  }
+
+  async blockUser (req, res) {
+    try {
+      const statusBlocked = await this.getStatusBlocked();
+      const user = await User.findByIdAndUpdate(
```
Comment (VasylLymych) [inline src/services/users.js]: What if user is already blocked or deleted?

File: src/services/users.js [hunk 30]
```diff
@@ -0,0 +1,165 @@
...
+        .limit(+query.take || 50)
+        .exec();
+
+      return users ? res.json(users) : [];
+    } catch (err) {
+      return res.boom.badRequest();
+    }
+  }
+
+  async updateUser (req, res) {
```
Comment (VasylLymych) [inline src/services/users.js:45]: What if user is already blocked or deleted?

File: src/services/users.js [hunk 31]
```diff
@@ -0,0 +1,165 @@
...
+
+  async getStatusDeleted (req, res) {
+    const statusDeleted = await UserStatuses.findOne({ name: 'Deleted' });
+    return statusDeleted;
+  }
+
+  async deleteUser (req, res) {
+    try {
+      const statusDeleted = await this.getStatusDeleted();
+      const user = await User.findByIdAndUpdate(
```
Comment (VasylLymych) [inline src/services/users.js]: What if user is already blocked or deleted?

File: src/services/users.js (hunk 27 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js (hunk 28 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js (hunk 29 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js (hunk 30 above)
Comment (asenhodzhev) [reply src/services/users.js:45]: fixed

File: src/services/users.js (hunk 31 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js [hunk 32]
```diff
@@ -0,0 +1,195 @@
...
+const { UserRoles } = require('../models');
+const { UserStatuses } = require('../models');
+const bcrypt = require('bcryptjs');
+
+class UsersService {
+  async updateUserPassword (req, res) {
+    try {
+      const { oldPassword, newPassword } = req.body;
+      const user = await User.findOne({ _id: req.params.id }).exec();
+
```
Comment (VasylLymych) [inline src/services/users.js]: Check if user exists here

File: src/services/users.js [hunk 33]
```diff
@@ -0,0 +1,195 @@
...
+            res.boom.badRequest('new password cannot be the same as old');
+          }
+        }
+      );
+    } catch (err) {
+      res.boom.notFound();
+    }
+  }
+
+  async getStatusActive () {
```
Comment (VasylLymych) [inline src/services/users.js]: Create one method "getStatus", which accepts status name

File: src/services/users.js [hunk 34]
```diff
@@ -0,0 +1,195 @@
...
+    const statusActive = await UserStatuses.findOne({ name: 'Active' });
+    return statusActive;
+  }
+
+  async getRoleUser () {
```
Comment (VasylLymych) [inline src/services/users.js]: Create one method "getRole", which accepts role name

File: src/services/users.js [hunk 35]
```diff
@@ -0,0 +1,195 @@
...
+    const userExist = await User.findOne({ email: req.body.email });
+    return userExist;
+  }
+
+  async createUser (req, res) {
+    try {
+      const roleUser = await this.getRoleUser();
+      const newUser = await this.createBasicUser(req);
+      newUser.role = roleUser._id;
+      const userExist = await this.checkUserExists(req);
```
Comment (VasylLymych) [inline src/services/users.js]: Check if user exists at start. There is no point to perform 2 calls to db (getStatus and getRole), if user already exists

File: src/services/users.js [hunk 36]
```diff
@@ -0,0 +1,195 @@
...
+        .select('-password')
+        .skip(+query.skip || 0)
+        .limit(+query.take || 50)
+        .exec();
+
+      const count = User.find(query).count();
+
+      return {
+        total_number: count,
+        allusers: users ? res.json(users) : res.json([])
```
Comment (VasylLymych) [inline src/services/users.js]: {total_size, list}

File: src/services/users.js (hunk 30 above)
Comment (VasylLymych) [reply src/services/users.js:45]: ?

File: src/routers/users.js [hunk 37]
```diff
@@ -1,9 +1,19 @@
...
-usersRouter.get('/:id', usersController.getUser);
+usersRouter.put('/:id', updateUserPayloadValidator, usersController.updateUser);
+usersRouter.get('/', usersController.getUsers);
+usersRouter.patch('/password/:id', usersController.updateUserPassword);
+usersRouter.post(
+  '/admin',
+  createUserPayloadValidator,
+  usersController.createAdmin
+);
+usersRouter.patch('/blocked/:id', usersController.blockUser);
```
Comment (VasylLymych) [inline src/routers/users.js]: /block/:id
delete(/:id)

File: src/services/users.js (hunk 30 above)
Comment (asenhodzhev) [reply src/services/users.js:45]: fixed

File: src/services/users.js (hunk 32 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js (hunk 33 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js (hunk 34 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js (hunk 35 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js (hunk 36 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/routers/users.js (hunk 37 above)
Comment (asenhodzhev) [reply src/routers/users.js]: fixed

File: src/services/users.js [hunk 38]
```diff
@@ -0,0 +1,191 @@
...
+        res.boom.badRequest('user already exists');
+      }
+    } catch (err) {
+      res.boom.badRequest();
+    }
+  }
+
+  async getUsers (req, res) {
+    const { _id, name, email, status, role, skip, take } =
+      (typeof req.query !== 'undefined' && req.query) || {};
```
Comment (VasylLymych) [inline src/services/users.js]: This check is useless here. There is always req.query object. It can just be empty

File: src/services/users.js [hunk 39]
```diff
@@ -0,0 +1,191 @@
...
+    const query = {};
+    if (_id != null) query._id = _id;
```
Comment (VasylLymych) [inline src/services/users.js]: No necessity to write if (_id != null)... You can just write if(_id)... Remove these nulls

File: src/services/users.js [hunk 40]
```diff
@@ -0,0 +1,191 @@
...
+    } catch {
+      res.boom.notFound();
+    }
+  }
+
+  async blockUser (req, res) {
+    try {
+      const statusBlocked = await this.getStatus('Blocked');
+      const currentStatus = await this.getCurrentStatus(req);
+      if (statusBlocked._id.toString() !== currentStatus.toString()) {
```
Comment (VasylLymych) [inline src/services/users.js]: Check here also if user is not deleted

File: src/services/users.js [hunk 41]
```diff
@@ -0,0 +1,191 @@
...
+        return newUser ? res.json(true) : res.boom.notFound();
+      } else {
+        res.boom.badRequest('user already exists');
+      }
+    } catch (err) {
+      res.boom.badRequest();
+    }
+  }
+
+  async createAdmin (req, res) {
```
Comment (VasylLymych) [inline src/services/users.js:11]: createUser and createAdmin almost the same methods. There is only difference in role. So move this all logic into mutual for them createUser method, which accepts body and role

File: src/services/users.js [hunk 42]
```diff
@@ -0,0 +1,191 @@
...
+  async createBasicUser (req, res) {
+    const statusActive = await this.getStatus('Active');
+    const newUser = new User(req.body);
+    newUser.status = statusActive._id;
+    const salt = await bcrypt.genSalt(10);
+    newUser.password = await bcrypt.hash(newUser.password, salt);
+    return newUser;
+  }
+
+  async checkUserExists (req, res) {
```
Comment (VasylLymych) [inline src/services/users.js]: Actually I don't think that we need these 4 methods (checkUserExists, getStatus, getRole, getCurrentStatus). They don't make code shorter

File: src/controllers/users.js [hunk 43]
```diff
@@ -1,16 +1,33 @@
-const { User } = require('../models');
+const { UsersService } = require('../services');
+const usersService = new UsersService();
```
Comment (VasylLymych) [inline src/controllers/users.js]: Make this field private, same as here https://github.com/Sf-62-NodeJS/clothes-market-server/blob/2dae079aa956d0614ab4171f7b618dede5ac8d01/src/controllers/sizes.js#L4

File: src/services/users.js (hunk 39 above)
Comment (asenhodzhev) [reply src/services/users.js]: Fixed

File: src/controllers/users.js (hunk 43 above)
Comment (asenhodzhev) [reply src/controllers/users.js]: Fixed

File: src/services/users.js (hunk 42 above)
Comment (asenhodzhev) [reply src/services/users.js]: Fixed

File: src/services/users.js (hunk 41 above)
Comment (asenhodzhev) [reply src/services/users.js:11]: Fixed

File: src/services/users.js (hunk 40 above)
Comment (asenhodzhev) [reply src/services/users.js]: Fixed

File: src/services/users.js (hunk 38 above)
Comment (asenhodzhev) [reply src/services/users.js]: Fixed

File: src/controllers/tests/users.test.js [hunk 44]
```diff
@@ -1,17 +1,113 @@
 const UsersController = require('../users');
+const { User } = require('../../models');
+const { UserStatuses } = require('../../models');
+// const { UserRoles } = require('../../models');
```
Comment (VasylLymych) [inline src/controllers/tests/users.test.js]: Remove this commented line

File: src/controllers/tests/users.test.js [hunk 45]
```diff
@@ -20,25 +116,151 @@ describe('usersController tests', () => {
...
     boom: {
       badRequest: jest.fn()
     },
-    json: payload => payload
+    json: (payload) => payload
   };
 
-  it('should return user', async () => {
-    const user = await usersController.getUser(requestStub, responseStub);
+  /* it('should return created user', async () => {
```
Comment (VasylLymych) [inline src/controllers/tests/users.test.js]: Remove commented tests. We don't need them

File: src/controllers/tests/users.test.js [hunk 46]
```diff
@@ -20,25 +116,151 @@ describe('usersController tests', () => {
...
+    };
+    const updatedUser = await usersController.updateUser(
+      requestStub,
+      responseStub
+    );
+
+    expect(updatedUser).toEqual(true);
+  });
+
+  /* it('should create admin', async () => {
```
Comment (VasylLymych) [inline src/controllers/tests/users.test.js]: Same as above

File: src/services/tests/users.test.js [hunk 47]
```diff
@@ -0,0 +1,271 @@
+const { UsersService } = require('..');
+const { User } = require('../../models');
+const { UserStatuses } = require('../../models');
+// const { UserRoles } = require('../../models');
```
Comment (VasylLymych) [inline src/services/tests/users.test.js]: Remove this commented line

File: src/services/tests/users.test.js [hunk 48]
```diff
@@ -0,0 +1,271 @@
...
+  };
+
+  const responseStub = {
+    boom: {
+      badRequest: jest.fn()
+    },
+    json: (payload) => payload
+  };
+
+  /* it('should create user', async () => {
```
Comment (VasylLymych) [inline src/services/tests/users.test.js]: Remove all commented tests

File: src/services/tests/users.test.js [hunk 49]
```diff
@@ -0,0 +1,271 @@
...
+          status: '12ad122xa7z'
+        })
+      };
+    };
+    const response = await usersService.updateUser(requestStub, responseStub);
+
+    expect(response).toEqual(true);
+  });
+
+  /* it('should create admin', async () => {
```
Comment (VasylLymych) [inline src/services/tests/users.test.js]: Same as above

File: src/services/users.js [hunk 50]
```diff
@@ -0,0 +1,178 @@
...
+      };
+    } catch (err) {
+      return res.boom.badRequest();
+    }
+  }
+
+  async updateUser (req, res) {
+    try {
+      const user = await User.findOne({ _id: req.params.id }).exec();
+      const currentStatus = user.status;
```
Comment (VasylLymych) [inline src/services/users.js:49]: Throw not found error here if user doesn't exist

File: src/services/users.js [hunk 51]
```diff
@@ -0,0 +1,178 @@
...
+      }
+    } catch (err) {
+      return res.boom.notFound();
+    }
+  }
+
+  async updateUserPassword (req, res) {
+    try {
+      const user = await User.findOne({ _id: req.params.id }).exec();
+      const currentStatus = user.status;
```
Comment (VasylLymych) [inline src/services/users.js:70]: Same as in comment above

File: src/services/users.js [hunk 52]
```diff
@@ -0,0 +1,178 @@
...
+        }
+      }
+    } catch {
+      res.boom.notFound();
+    }
+  }
+
+  async blockUser (req, res) {
+    try {
+      const statusBlocked = await UserStatuses.findOne({
```
Comment (VasylLymych) [inline src/services/users.js]: You can get all statuses instead of two separate calls

File: src/services/users.js [hunk 53]
```diff
@@ -0,0 +1,178 @@
...
+        name: 'Blocked'
+      }).exec();
+      const statusDeleted = await UserStatuses.findOne({
+        name: 'Deleted'
+      }).exec();
+      const user = await User.findOne({ _id: req.params.id }).exec();
+      const currentStatus = user.status;
```
Comment (VasylLymych) [inline src/services/users.js:109]: Throw not found error here if user doesn't exist

File: src/services/users.js [hunk 54]
```diff
@@ -0,0 +1,178 @@
...
+    }
+  }
+
+  async deleteUser (req, res) {
+    try {
+      const statusDeleted = await UserStatuses.findOne({
+        name: 'Deleted'
+      }).exec();
+      const user = await User.findOne({ _id: req.params.id }).exec();
+      const currentStatus = user.status;
```
Comment (VasylLymych) [inline src/services/users.js:137]: Throw not found error here if user doesn't exist

File: src/services/users.js [hunk 55]
```diff
@@ -0,0 +1,178 @@
...
+  async #createBaseUser (req, res, userRole) {
+    try {
+      const statusActive = await UserStatuses.findOne({
+        name: 'Active'
+      }).exec();
+      const newUser = new User(req.body);
+      newUser.status = statusActive._id;
+      const salt = await bcrypt.genSalt(10);
+      newUser.password = await bcrypt.hash(newUser.password, salt);
+      const userExist = User.findOne({ email: req.body.email });
```
Comment (VasylLymych) [inline src/services/users.js]: check if user exists at start. If user exists, but his status is deleted, we can create new with such an email

File: src/controllers/tests/users.test.js (hunk 44 above)
Comment (asenhodzhev) [reply src/controllers/tests/users.test.js]: fixed

File: src/controllers/tests/users.test.js (hunk 45 above)
Comment (asenhodzhev) [reply src/controllers/tests/users.test.js]: fixed

File: src/controllers/tests/users.test.js (hunk 46 above)
Comment (asenhodzhev) [reply src/controllers/tests/users.test.js]: fixed

File: src/services/tests/users.test.js (hunk 47 above)
Comment (asenhodzhev) [reply src/services/tests/users.test.js]: fixed

File: src/services/tests/users.test.js (hunk 48 above)
Comment (asenhodzhev) [reply src/services/tests/users.test.js]: fixed

File: src/services/tests/users.test.js (hunk 49 above)
Comment (asenhodzhev) [reply src/services/tests/users.test.js]: fixed

File: src/services/users.js (hunk 50 above)
Comment (asenhodzhev) [reply src/services/users.js:49]: fixed

File: src/services/users.js (hunk 51 above)
Comment (asenhodzhev) [reply src/services/users.js:70]: fixed

File: src/services/users.js (hunk 53 above)
Comment (asenhodzhev) [reply src/services/users.js:109]: fixed

File: src/services/users.js (hunk 54 above)
Comment (asenhodzhev) [reply src/services/users.js:137]: fixed

File: src/services/users.js (hunk 55 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js (hunk 52 above)
Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/services/users.js [hunk 56]
```diff
@@ -0,0 +1,186 @@
...
+      if (currentStatus.toString() === statusActive._id.toString()) {
+        const updated = await User.findByIdAndUpdate(req.params.id, req.body, {
+          new: true
+        }).exec();
+        return updated ? res.json(true) : res.boom.notFound();
+      } else {
+        res.boom.badRequest('user is not active');
+      }
+    } catch (err) {
+      return res.boom.notFound();
```
Comment (VasylLymych) [inline src/services/users.js]: You already throw notFound error. So here surely will not be not found error. Throw here badRequest. Fix it everywhere

Comment (asenhodzhev) [reply src/services/users.js]: fixed

File: src/routers/users.js [hunk 57]
```diff
@@ -1,9 +1,19 @@
...
 const { UsersController } = require('../controllers');
 const { createUserPayloadValidator } = require('../middlewares/validators');
+const { updateUserPayloadValidator } = require('../middlewares/validators');
 
 const usersController = new UsersController();
 usersRouter.post('/', createUserPayloadValidator, usersController.createUser);
-usersRouter.get('/:id', usersController.getUser);
+usersRouter.put('/:id', updateUserPayloadValidator, usersController.updateUser);
+usersRouter.get('/', usersController.getUsers);
+usersRouter.patch('/password/:id', usersController.updateUserPassword);
```
Comment (vhazna) [inline src/routers/users.js]: You need a password validation here.

Comment (asenhodzhev) [reply src/routers/users.js]: Fixed

File: src/services/users.js [hunk 58]
```diff
@@ -0,0 +1,186 @@
...
+        newUser.password = await bcrypt.hash(newUser.password, salt);
+        const roleUser = await UserRoles.findOne({ name: userRole }).exec();
+        newUser.role = roleUser._id;
+        await newUser.save();
+        return newUser ? res.json(true) : res.boom.notFound();
+      } else {
+        res.boom.badRequest('user already exists');
+      }
+    } catch (err) {
+      res.boom.badRequest();
```
Comment (VasylLymych) [inline src/services/users.js]: You need to throw here internal server error instead of badRequest. Fix this everywhere before merging

**Review by VasylLymych — APPROVED** [review]
